src/transaction_queue.rs

```diff
@@ -1,35 +1,107 @@
-use std::cmp;
 use std::collections::{HashSet, VecDeque};
+use std::{cmp, fmt};
 
-use rand;
+use rand::{self, Isaac64Rng, Rng};
 
 use traits::Contribution;
+use util::SubRng;
+
+/// An interface to the transaction queue. A transaction queue is a structural part of
+/// `QueueingHoneyBadger` that manages enqueueing of transactions for a future batch and dequeueing
+/// of transactions to become part of a current batch.
+pub trait TransactionQueue<T> {
+    fn is_empty(&self) -> bool;
+    fn push_back(&mut self, t: T);
+    fn choose(&mut self, amount: usize, batch_size: usize) -> Vec<T>;
+    fn remove_all<'a, I>(&mut self, txs: I)
+    where
+        I: IntoIterator<Item = &'a T>,
+        T: 'a + Contribution;
+}
 
 /// A wrapper providing a few convenience methods for a queue of pending transactions.
-#[derive(Debug)]
-pub struct TransactionQueue<T>(pub VecDeque<T>);
+pub struct VecDequeTransactionQueue<T> {
+    /// Random number generator passed on from the algorithm instance.
```

Comment:
I'd document what it's used for rather than where it happens to come from in the caller. Maybe: "Random number generator used for choosing transactions from the queue"?

Alternatively, we could not even store it in here, but just make it an argument of `choose`?

Reply:
In some queues `choose` could be a deterministic choice, in others it could use a thread-local RNG, so that passing it between threads wouldn't be an issue at the trait level. So I'm leaning towards keeping `rng` internal to a particular queue implementation.